### PR TITLE
Docs: Added bson section

### PR DESCRIPTION
**Pull request by rozza**
Updated scaladoc
SCALA-197

[Review on Reviewable](https://reviewable.io/reviews/rozza/mongo-scala-driver/9)

### CODE REVIEW

**Comment by rozza** [comment]
Docs: http://rozza.github.io/mongo-scala-driver/1.0/bson/
Tools examples: https://github.com/rozza/mongo-scala-driver/tree/s197/examples/src/test/scala/tools

**Comment by rozza** [comment]
Removed the static api docs from the PR as it made it too unwieldy

File: docs/reference/content/bson/documents.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+The scala driver includes two scala specific representations for BSON documents. Following convention from the scala collections library,
+there are immutable and mutable implementations of documents. The underlying implementations of the scala Document use the type safe 
+[`BsonDocument`]({{< coreapiref "org/bson/BsonDocument" >}}) class. The scala bson classes are available from the `org.mongodb.scala.bson` 
+namespace, which includes type aliases and companion objects.  In general this should suffice but for advanced use cases you may need to 
+use `org.bson` directly.
+
+{{% note %}}
+The scala `Document` classes implement `TraversableLike[(String, BsonValue)]` and the general API mirrors that of a `Map[String, BsonValue]`.
+However, unlike `Map` implementations of `TraversableLike` enables strict type safety as their is no variance in the value type.
```
Comment (jyemin) [inline docs/reference/content/bson/documents.md]: "as there is"

File: docs/reference/content/bson/documents.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+{{% /note %}}
+
+[`BsonValue`]({{< apiref "org.mongodb.scala.bson.BsonValue" >}}) is the type safe representation of a Bson type from the `org.bson` library,
```
Comment (jyemin) [inline docs/reference/content/bson/documents.md]: Run-on sentence

File: docs/reference/content/bson/documents.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+| Boolean   | `Boolean`                         |
+| Double    | `Double`                          |
+| Int32     | `Integer`                         |
+| Int64     | `Long`                            |
+| String    | `String`                          |
+| Binary    | `Array[Byte]`                     |
+| ObjectId  | `ObjectId`                        |
+| Null      | `None`                            |
+
+It is actually possible to change or extend these mappings this will be covered in detail below.
```
Comment (jyemin) [inline docs/reference/content/bson/documents.md]: Run-on sentence

File: docs/reference/content/bson/documents.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+|-----------|-----------------------------------|
+| Document  | `org.mongodb.scala.bson.Document` |
+| Array     | `List`                            |
+| Date      | `Date` or `int` (ms since epoch)  |
+| Boolean   | `Boolean`                         |
+| Double    | `Double`                          |
+| Int32     | `Integer`                         |
+| Int64     | `Long`                            |
+| String    | `String`                          |
+| Binary    | `Array[Byte]`                     |
```
Comment (jyemin) [inline docs/reference/content/bson/documents.md:35]: what if you need to specify the binary sub-type?

File: docs/reference/content/bson/documents.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+and `org.mongodb.scala.bson.Document` as well as being available from `org.mongodb.scala.bson.collection.immutable.Document`. Instances of 
+this type are guaranteed to be immutable for everyone. Such a collection will never change after it is created. Therefore, you can rely on 
+the fact that accessing the same collection value repeatedly at different points in time will always yield a collection with the same elements.
+
+```scala
+import org.mongodb.scala.bson._
+
+val doc1 = Document("AL" -> BsonString("Alabama"))
+val doc2 = doc1 + ("AK" -> BsonString("Alaska"))
+val doc3 = doc2 ++ Document("AR" -> BsonString("Arkansas"), "AZ" -> BsonString("Arizona"))
```
Comment (jyemin) [inline docs/reference/content/bson/documents.md:55]: Maybe just for my own edification, but can you show an example of an immutable array as well?

File: docs/reference/content/bson/documents.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+val doc = Document("AL" -> BsonString("Alabama"))
+val doc1 = doc + ("AK" -> BsonString("Alaska"))   // doc not mutated but new doc created
+doc1 ++= Document("AR" -> BsonString("Arkansas"), 
+                  "AZ" -> BsonString("Arizona"))  // doc1 mutated as ++= changes in place. 
+```
+
+## Implicit conversions
+
+For many of the `BsonValue` types there are obvious direct mappings from a Scala type. For example, a `String` maps to `BsonString`, an `Int`
+maps to `BsonInt32` and a `Long` maps to a `BsonInt64`.  For convenience these types can be used directly with `Documents` and, they are
```
Comment (jyemin) [inline docs/reference/content/bson/documents.md]: extraneous comma between "and" and "they"

File: docs/reference/content/bson/documents.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+> This parameter is called the magnet. Its type is the magnet type, a dedicated type constructed purely as the target of a number of implicit conversions defined in the magnets companion object, which are called the magnet branches and which model the various “overloads”.
+
+Source: [The Magnet Pattern](http://spray.io/blog/2012-12-13-the-magnet-pattern/)
+
+
+In the API where we would normally expect a single value or a key value pair or many key value pairs eg: (`BsonValue`, `(String, BsonValue)` 
+or `Iterable[(String, BsonValue)]`) we require anything that can become those types via _"`CanBeX`"_ traits that handle the implicit 
+conversions necessary to conform to the correct types. These traits are [`CanBeBsonValue`]({{< apiref "org.mongodb.scala.bson.BsonMagnets$$CanBeBsonValue" >}}), 
+[`CanBeBsonElement`]({{< apiref "org.mongodb.scala.bson.BsonMagnets$$CanBeBsonElement" >}}) and
+[`CanBeBsonElements`]({{< apiref "org.mongodb.scala.bson.BsonMagnets$$CanBeBsonElements" >}}).
```
Comment (jyemin) [inline docs/reference/content/bson/documents.md:127]: Maybe an example?

File: examples/src/test/scala/tools/mongoexport.scala [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+      case Some(value) => cursor.limit(value)
+    }
+    options.sort match {
+      case None => cursor
+      case Some(value) => cursor.sort(value)
+    }
+
+    output.write("")
+    cursor.subscribe(
+      (doc: Document) => output.write(s"${doc.toJson()}\r\n"),
```
Comment (jyemin) [inline examples/src/test/scala/tools/mongoexport.scala]: String.format in Java allows use of %n for platform-independent line endings.

File: examples/src/test/scala/tools/mongoimport.scala [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+   * @param collection the mongodb collection to insert into
+   * @param lines the iterator from the importSource
+   * @param promise the promise that is fufilled on completion or on error
+   * @return the promise
+   */
+  private def importJson(collection: MongoCollection[Document], lines: Iterator[String], promise: Promise[Completed]): Promise[Completed] = {
+    lines.hasNext match {
+      case true => collection.insertMany(lines.take(1000).map(json => Document(json)).toSeq).subscribe(
+        (completed: Completed) => lines.hasNext match {
+            case true => importJson(collection, lines, promise)
```
Comment (jyemin) [inline examples/src/test/scala/tools/mongoimport.scala]: The docs for Iterator.take say:

```
 Reuse: After calling this method, one should discard the iterator it was called on, and use only the iterator that was returned. Using the old iterator is undefined, subject to change, and may result in changes to the new iterator as well
```

Not sure if that's an issue here or not.

**Comment by jyemin** [comment]
Finished initial review.  Looks good.

File: docs/reference/content/bson/documents.md (hunk 1 above)
Comment (rozza) [inline docs/reference/content/bson/documents.md]: Done

File: docs/reference/content/bson/documents.md (hunk 2 above)
Comment (rozza) [inline docs/reference/content/bson/documents.md]: Done

File: docs/reference/content/bson/documents.md (hunk 3 above)
Comment (rozza) [inline docs/reference/content/bson/documents.md]: Done

File: docs/reference/content/bson/documents.md (hunk 4 above)
Comment (rozza) [inline docs/reference/content/bson/documents.md:35]: A `BsonBinary` would have to be used.

File: docs/reference/content/bson/documents.md (hunk 5 above)
Comment (rozza) [inline docs/reference/content/bson/documents.md:55]: Not sure I follow - an immutable array of what, to do what?

File: docs/reference/content/bson/documents.md (hunk 6 above)
Comment (rozza) [inline docs/reference/content/bson/documents.md]: Done

File: docs/reference/content/bson/documents.md (hunk 7 above)
Comment (rozza) [inline docs/reference/content/bson/documents.md:127]: Done.

File: examples/src/test/scala/tools/mongoexport.scala (hunk 8 above)
Comment (rozza) [inline examples/src/test/scala/tools/mongoexport.scala]: ah theres a println - to do it. Updated

File: examples/src/test/scala/tools/mongoimport.scala (hunk 9 above)
Comment (rozza) [inline examples/src/test/scala/tools/mongoimport.scala]: Good catch

**Comment by rozza** [comment]
Updated

File: docs/reference/content/bson/documents.md (hunk 5 above)
Comment (jyemin) [inline docs/reference/content/bson/documents.md:55]: I mean an example of how to add an array of BSON values, e.g.

   val docs4 = docs3 ++ List(BsonInt32(1), BsonInt32(5))

**Comment by jyemin** [comment]
Re-reviewed.  Just the one comment on the second commit.

File: docs/reference/content/bson/documents.md (hunk 5 above)
Comment (rozza) [inline docs/reference/content/bson/documents.md:55]: You can't as it has to be key value pairs - docs3 is essentially a map and it doesn't make sense to union a map and a list.  Adding a traversable of `[(String, BsonValue)]` pairs is already shown in: 

`doc2 ++ Document("AR" -> BsonString("Arkansas"), "AZ" -> BsonString("Arizona"))`

**Comment by rozza** [comment]
Updated, rebased and merged

**Pull request closed without merging**